### DISCUSSION TITLE
Machine Pin internal pull resistor mode using integer

### DISCUSSION THREAD

**Discussion by puppet13th** [micropython/micropython#10269] [Hardware & Peripherals] [answered]
We can uses Pin.PULL_UP to enable internal pull up resistor and Pin.PULL_DOWN to enable pull down resistor.

eg.
```py
from machine import Pin
pin = Pin( 0, Pin.IN, Pin.PULL_UP ) # pull up
pin = Pin( 0, Pin.IN, Pin.PULL_DOWN ) # pull down
```

we also can use integer for internal resistor pull mode
eg
```py
from machine import Pin
pin = Pin( 0, Pin.IN, 2 ) # pull up
pin = Pin( 0, Pin.IN, 1) # pull down
pin = Pin( 0, Pin.IN, 0 ) # disable internal pull resistor
```

i have tried to use integer value 3 as pull resistor mode using these code:
```py
from machine import Pin, ADC
from time import sleep

adc = ADC( Pin( 0 ) )
adc.atten(ADC.ATTN_11DB)
pin = Pin( 0, Pin.IN )

for num in range (-12, 12):
    pin.init( pull=0 ) # reset
    sleep( 0.5 )
    pin.init( pull=num )
    adc_list = []
    pin_list = []
    for loop in range( 0, 20 ):
        adc_list.append( adc.read_u16() )
        pin_list.append( pin() )
        sleep( 0.2 )
    adc_mean = sum( adc_list ) / 20
    pin_mean = sum (pin_list ) / 20
    print( num, adc_mean, pin_mean )
```
edit : updated test code

when nothing connected to pin 0
adc.read_u16 always return half of full value (33k ish,max value = 65k)
pin() always return 0

what does pull=-1 and pull=3 mode do?
pull=-1 --> ???
pull=0 --> disable pull resistor
pull=1 --> pull down
pull=2 --> pull up
pull=3 --> ???

I'm using mpy 1.19.1 October nightly + lolin c3 mini board(esp32-c3)

**Comment by peterhinch** [comment, marked as answer]
My guess is that it's enabling both a pull up and a pull down.

**Reply by puppet13th** [reply]
after further test, i can confirm mode 3 is both pull up and pull down resistor enabled.

**Reply by puppet13th** [reply]
another correction
mode -1 = mode 3 = both pull up and down
there is same pattern for pull up mode using integer.
-1 -> 0 -> 1 -> 2 -> 3
both -> disabled -> down -> up -> both

**Comment by puppet13th** [comment]
wiring:
3v3 -- Resistor 39k -- Net1 -- resistor 10k -- GND
Net1 -- switch -- gpio0

test result:
![mpy pin pull mode](https://user-images.githubusercontent.com/48564964/208421766-3ba5d3ea-cd90-4c49-bd1c-284eff81e128.png)

**Reply by peterhinch** [reply]
There is no mystery here: it's just binary. Bit 0 enables pull down and bit 1 enables pull up. So, for example, -1 is 1111...11111 in binary. Bits 0 and 1 are both set so both pull resistors get enabled. To take another example, 6 is 110 in binary. Bit 0 is off, so no pull down. Bit 1 is on, so we have a pull up.
